// @fitbit/sdk-cli/lib/auth/ResourceOwnerPasswordRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ResoureOwnerTokenRequest {
    constructor(request) {
        this.clientId = request.client_id;
        this.username = request.username;
        this.password = request.password;
        this.extras = request.extras;
    }
    toJson() {
        return {
            grant_type: 'password',
            username: this.username,
            password: this.password,
            client_id: this.clientId,
            extras: this.extras,
        };
    }
    toStringMap() {
        const map = {
            grant_type: 'password',
            client_id: this.clientId,
            username: this.username,
            password: this.password,
        };
        if (this.extras) {
            for (const extra in this.extras) {
                if (this.extras.hasOwnProperty(extra) && !map.hasOwnProperty(extra)) {
                    map[extra] = this.extras[extra];
                }
            }
        }
        return map;
    }
}
exports.default = ResoureOwnerTokenRequest;
